master/internal/command/events.go

```diff
@@ -44,17 +48,70 @@ type event struct {
 	LogEvent *string `json:"log_event"`
 }
 
+type logSubscribers = map[*actor.Ref]webAPI.LogsRequest // CHECK could we get different pointers to
+// the same ref? thus resulting in different map keys
+
+// GetEventCount is an actor message used to get the number of events in buffer.
+type GetEventCount struct{}
+
 type eventManager struct {
-	bufferSize int
-	buffer     *ring.Ring
-	closed     bool
-	seq        int
+	bufferSize   int
+	buffer       *ring.Ring
+	closed       bool
+	seq          int
+	isTerminated bool // DISCUSS If we don't want to don't keep track of this here we'd either need to
+	// expose event externally or define and send a new message. also the logActor might not receive
+	// the termination log message (based on LogRequest) if we rely on exposing Event..
+	logStreams logSubscribers
 }
 
 func newEventManager() *eventManager {
 	return &eventManager{
-		bufferSize: defaultEventBufferSize,
-		buffer:     ring.New(defaultEventBufferSize),
+		bufferSize:   defaultEventBufferSize,
+		buffer:       ring.New(defaultEventBufferSize),
+		logStreams:   make(logSubscribers),
+		isTerminated: false,
+	}
+}
+
+// QUESTION where do we house these utilities?
+func countNonNullRingValues(ring *ring.Ring) int {
+	// OPT we could use log_buffer here instead of a plain ring buffer.
+	count := 0
+	ring.Do(func(val interface{}) {
+		if val != nil {
+			count++
+		}
+	})
+	return count
+}
+
+// TODO inline this.
+func (e *eventManager) RemoveSusbscribers(ctx *actor.Context) {
+	for actor := range e.logStreams {
+		// OPT this will trigger a bunch of CloseStream message that'll come back to eventManager.
+		ctx.Tell(actor, webAPI.CloseStream{})
+	}
+	// REMOVE ME
+	e.logStreams = make(logSubscribers)
```

Comment:
What's this for?  Could you do `e.logStreams = nil`?

Reply:
huh, didn't know nil would still be a valid slice :+1: